cx/spark/server.js

```diff
@@ -0,0 +1,199 @@
+ */
+const client = new SessionsClient({apiEndpoint: locationId + '-dialogflow.googleapis.com'});
+
+//Upon start a webhook is registered with spark
+//Upon closure the webhook is removed from spark
+
+const listener = app.listen(process.env.PORT, async function() {
+  await init();
+  console.log('Your Spark integration server is listening on port '
+      + listener.address().port);
+});
+
+app.post('/', async function(req, res) {
+  const message = await retrieveMessage(req.body.data.id);
+  const dialogflowResponse = await detectIntentText(message.text, req.body.data.personId);
+  sendMessage(dialogflowResponse, message.email);
+});
+
+process.on('SIGTERM', () => {
+  listener.close(async ()=>{
+    console.log('Closing http server.');
+    await deleteWebhooks();
+    process.exit(0);
+  });
+});
+
+async function init(){
+  await deleteWebhooks();
+  registerWebhook();
+}
+
+/**
+ * This function calls Dialogflow CX API to retrieve the response
+ * https://cloud.google.com/dialogflow/cx/docs/quick/api
+ */
+
+async function detectIntentText(query,personId) {
+    let agentResponse ='';
+
+  const sessionId = personId;
+  const sessionPath = client.projectLocationAgentSessionPath(
+      projectId,
+      locationId,
+      agentId,
+      sessionId
+  );
+  console.info(sessionPath);
+
+  const request = {
+    session: sessionPath,
+    queryInput: {
+      text: {
+        text: query,
+      },
+      languageCode,
+    },
+  };
+  const [response] = await client.detectIntent(request);
+  console.log(`User Query: ${query}`);
+  for (const message of response.queryResult.responseMessages) {
+    if (message.text) {
+      console.log(`Agent Response: ${message.text.text}`);
+      agentResponse += `${message.text.text}\n`;
+    }
+    
+  }
+  if (response.queryResult.match.intent) {
+    console.log(
+        `Matched Intent: ${response.queryResult.match.intent.displayName}`
+    );
+  }
+  console.log(
+      `Current Page: ${response.queryResult.currentPage.displayName}`
+  );
+
+  return agentResponse;
+}
+
+function sendMessage(text, personEmail) {
+  request.post('https://api.ciscospark.com/v1/messages', {
+    auth: {
+      bearer: sparkAccessToken
+    },
+    json: {
+      "toPersonEmail": personEmail,
+      "text": text
+    }
+  }, (err, resp, body) => {
+    if (err) {
+      console.error('Failed to send message :' + err);
+    }
+  });
+}
+
+function registerWebhook() {
+  request.post('https://api.ciscospark.com/v1/webhooks', {
+    auth: {
+      bearer: sparkAccessToken
+    },
+    json: {
+      "name": "test",
+      "targetUrl": targetUrl,
+      "resource": "messages",
+      "event": "created"
+    }
+  }, (err, resp, body) => {
+    if (err) {
+      console.error('Failed to create Webhook :' + err);
+    }
+  });
+}
+
+function deleteWebhooks() {
+  return new Promise((resolve, reject) =>{
+    request.get('https://api.ciscospark.com/v1/webhooks?max=100', {
```

Comment:
Looks like we are trying to list all the webhooks and then do the match. Is there any GET method with parameter we can use directly?

Reply:
I don't see any change in this line? Could you please clarify it is because we can only do this with LIST?